ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -805,19 +861,39 @@ void handleBody(InputStream payload)
               MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent, blobIdFactory);
           chunkSize = compositeBlobInfo.getChunkSize();
           totalSize = compositeBlobInfo.getTotalSize();
-          List<StoreKey> keys = compositeBlobInfo.getKeys();
-          chunkIdIterator = keys.listIterator();
-          numChunksTotal = keys.size();
-          dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
-          for (int i = 0; i < dataChunks.length; i++) {
-            dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+          try {
+            List<StoreKey> keys = compositeBlobInfo.getKeys();
+            if (hasRange) {
+              validatedByteRange = new ValidatedByteRange(options.getRange(), totalSize);
+              // Get only the chunks within the range.
+              firstChunkIndexInRange = (int) (validatedByteRange.getStartOffset() / chunkSize);
+              int lastChunkIndexInByteRange = (int) (validatedByteRange.getEndOffset() / chunkSize);
```

Comment:
this is the (inclusive) index of the last chunk in the byte range, but List.subList takes in an exclusive end index. should i change this to be exclusive?